torchrl/data/rlhf/utils.py

```diff
@@ -0,0 +1,391 @@
+        score_clip (float, optional): Scores from the reward model are clipped to the
+            range ``(-score_clip, score_clip)``. Defaults to 10.
+
+    Examples:
+        >>> from tensordict.nn import TensorDictModule
+        >>> from torchrl.modules.models.rlhf import GPT2RewardModel
+        >>> from torchrl.data.rlhf.utils import RolloutFromModel
+        >>> from torchrl.data.rlhf.dataset import get_dataloader
+        >>> from torchrl.data.rlhf.prompt import PromptData
+        >>> from transformers import GPT2LMHeadModel
+        >>>
+        >>> dl = get_dataloader(
+        ...     batch_size=4,
+        ...     block_size=550,
+        ...     tensorclass_type=PromptData,
+        ...     device="cpu",
+        ...     dataset_name="CarperAI/openai_summarize_tldr",
+        ... )
+        >>> model = GPT2LMHeadModel.from_pretrained("gpt2")
+        >>> # we load ref_model with random weights so it differs from model
+        >>> ref_model = GPT2LMHeadModel(GPT2LMHeadModel.config_class())
+        >>> reward_model = GPT2RewardModel(model_path="gpt2")
+        >>> rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+        >>>
+        >>> batch = next(dl)
+        >>> rollout = rollout_from_model.rollout_from_data(batch)
+        >>> rollout
+        TensorDict(
+            fields={
+                action: Tensor(shape=torch.Size([4, 50]), device=cpu, dtype=torch.int64, is_shared=False),
+                attention_mask: Tensor(shape=torch.Size([4, 50, 600]), device=cpu, dtype=torch.bool, is_shared=False),
+                input_ids: Tensor(shape=torch.Size([4, 50, 600]), device=cpu, dtype=torch.int64, is_shared=False),
+                next: TensorDict(
+                    fields={
+                        attention_mask: Tensor(shape=torch.Size([4, 50, 600]), device=cpu, dtype=torch.bool, is_shared=False),
+                        done: Tensor(shape=torch.Size([4, 50, 1]), device=cpu, dtype=torch.bool, is_shared=False),
+                        input_ids: Tensor(shape=torch.Size([4, 50, 600]), device=cpu, dtype=torch.int64, is_shared=False),
+                        reward: Tensor(shape=torch.Size([4, 50, 1]), device=cpu, dtype=torch.float32, is_shared=False),
+                        reward_kl: Tensor(shape=torch.Size([4, 50, 1]), device=cpu, dtype=torch.float32, is_shared=False),
+                        reward_raw: Tensor(shape=torch.Size([4, 50, 1]), device=cpu, dtype=torch.float32, is_shared=False)},
+                    batch_size=torch.Size([4, 50]),
+                    device=cpu,
+                    is_shared=False),
+                sample_log_prob: Tensor(shape=torch.Size([4, 50, 1]), device=cpu, dtype=torch.float32, is_shared=False)},
+            batch_size=torch.Size([4, 50]),
+            device=cpu,
+            is_shared=False)
+    """
+
+    EOS_TOKEN_ID = 50256
+
+    def __init__(
+        self, model, ref_model, reward_model, max_new_tokens=50, score_clip=10.0
```

Comment:
long term we may want to make this reward model optional and leave to the user the option of computing reward at a different time (eg when populating the replay buffer)